### PR TITLE
[feature] (planner) InferPredicate (#7096)

### LINKED ISSUE #7096: [Feature] Transitive Closure
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Transitive Closure 应用在 Doris 中主要指的是添加冗余 SQL 连接条件，通过冗余条件减少查询的数据量从而提升查询性能的一种 SQL 优化。 
形式上，这称为传递闭包，即如果 a=b 和 a=1，则它遵循 b=1。

Doris 当前支持两个表之间的谓词传递，比如 a=b and a=1 。但他不支持多个表之间的传递。比如 a= b and b=c and a=1 这时，a=1 只能传递为 b=1 但是不能传递出 c=1。

### Use case

举个例子
```
select * from t1, t2, t3 where t1.id=t2.id and t2.id=t3.id and t3.id=1;
```
上面这个查询他应该能传递出下面两个谓词
```
t2.id = 1
t1.id = 1
```
则查询效果应该等同于
```
select * from t1, t2, t3 where t1.id=t2.id and t2.id=t3.id and t3.id=1 and t2.id=1 and t1.id=1;
```

这么做的好处在于当 filter predicate 传递给了其他表，他就可以提前过滤掉大量数据，从而提升查询性能。

### Detail design

当前代码实现位置：PredicatePushDown.java

优化后位置：语义解析完成后，统一对 Predicate 进行扩展，格式化，去重及合并。

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhengshiJ**
## Proposed changes

this pr is for #7096  , which is add a function for infer predicate. 

- Add a switch enable_infer_predicate to control whether to perform predicate expansion.
- Register a new rule InferFiltersrule and add it to GlobalState.
- Traverse Conjunct to construct on/where equivalence connection, numerical connection and isNullPredicate.
- Infer all equivalence connections
- Construct additional numerical connections and isNullPredicate

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by EmmyMiao87** [comment]
Please change your PR title and msg by https://doris.apache.org/master/zh-CN/developer-guide/commit-format-specification.html#commit-%E6%A0%BC%E5%BC%8F%E8%A7%84%E8%8C%83

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+                                       ArrayList<Expr> newPredicatesList,
+                                       ArrayList<Expr> tmpSlotToLiteralPredicatesList,
+                                       boolean needAddNewPredicatesList) {
+        if (expr instanceof BinaryPredicate) {
+            BinaryPredicate newBP = (BinaryPredicate) expr;
+            Pair<Expr, Expr> pair = new Pair<>(newBP.getChild(0), newBP.getChild(1));
+            if (!slotToLiteralPredicatesSet.contains(pair)) {
+                slotToLiteralPredicatesSet.add(pair);
+                if (needAddNewPredicatesList) {
+                    newPredicatesList.add(newBP);
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:232]: Should the new predicate also be de-duplicated?
for example
a=b and b=c and a=1 and c=1
The b=1 may be inferred twice.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+            for (int index = 0; index < newPredicatesList.size(); ++index) {
+                Expr slotToLiteral = newPredicatesList.get(index);
+                buildNewBinaryPredicate(slotToLiteral, slotEqSlotPredicatesList, slotToLiteralPredicatesSet,
+                        newPredicatesList, tmpSlotToLiteralPredicatesList, analyzer);
+            }
+
+            int endSize = tmpSlotToLiteralPredicatesList.size();
+            tmpSize = endSize - startSize;
+
+        } while (tmpSize != 0);
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:161]: If there is no new tmp predicate but new predicate, could new predicate infer the new tmp predicate.
My suggestion is following:
Do not distinguish between tmp and new predicate in the process, just as a mark.
Just don't bring tmp when constructing a new expr at the end.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+            int startSize = tmpSlotToLiteralPredicatesList.size();
+
+            for (int index = 0; index < tmpSlotToLiteralPredicatesList.size(); ++index) {
+                Expr slotToLiteral = tmpSlotToLiteralPredicatesList.get(index);
+                buildNewBinaryPredicate(slotToLiteral, slotEqSlotPredicatesList, slotToLiteralPredicatesSet,
+                        newPredicatesList, tmpSlotToLiteralPredicatesList, analyzer);
+            }
+
+            for (int index = 0; index < newPredicatesList.size(); ++index) {
+                Expr slotToLiteral = newPredicatesList.get(index);
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:153]: Although this writing is functionally correct, there may be a risk of infinite loop if the logic is wrong.
Is it better to derive all the expressions of a single slotToLiteral predicate at once?

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+                    isNullPredicatesList, isNullPredicatesSet);
+
+            inferSlotPredicates(slotEqSlotPredicatesList, slotToLiteralPredicatesSet,
+                    slotToLiteralPredicatesList, newPredicatesList, analyzer);
+
+            inferIsNullPredicates(slotEqSlotPredicatesList, isNullPredicatesList,
+                    isNullPredicatesSet, newPredicatesList, analyzer);
+            
+            if (!newPredicatesList.isEmpty()) {
+                Expr rewriteExpr = new CompoundPredicate(CompoundPredicate.Operator.AND, expr, newPredicatesList.get(0));
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:81]: done

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+    private void initAllStructure(Expr conjunct,
+                                  ArrayList<Expr> slotEqSlotPredicatesList,
+                                  Set<Pair<Expr, Expr>> slotEqSlotPredicatesSet,
+                                  ArrayList<Expr> slotToLiteralPredicatesList,
+                                  Set<Pair<Expr, Expr>> slotToLiteralPredicatesSet,
+                                  ArrayList<Expr> isNullPredicatesList,
+                                  Set<Expr> isNullPredicatesSet) {
+        if (conjunct instanceof CompoundPredicate) {
+            for (int index = 0; index < conjunct.getChildren().size(); ++index) {
```
[the remark this answers is not in the document]
Comment (EmmyMiao87) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:123]: else return ?

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+    }
+
+    /**
+     * To determine whether it needs to be extended.
+     * eg:t1.id = t2.id and t2.id = 1;
+     * @param newSlot: t1.id
+     * @param checkSlot: t2.id
+     * @return needInfer.    True: needInfer. False: not needInfer
+     */
+    private boolean isNeedInfer(SlotRef newSlot, SlotRef checkSlot) {
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:312]: FROM t0 left outer join t1 inner join t2 where t1.id=t2.id and t2.id=1
jop = left outer join
checkjop = inner join
But it still can infer , right?

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+        List<Expr> isNullExpr =
+                (clauseType == ExprRewriter.ClauseType.ONCLAUSE) ? analyzer.getOnIsNullExpr() : new ArrayList<>();
+
+        //isNullDeDuplication: De-Duplication for isNullExpr
+        Set<Expr> isNullDeDuplication =
+                (clauseType == ExprRewriter.ClauseType.ONCLAUSE) ? analyzer.getOnIsNullDeDuplication() : Sets.newHashSet();
+
+
+        // exprToInteger/integerToExpr: function is easy to build warshall and newExprWithState
+        Map<Expr, Integer> exprToInteger = new HashMap<>();
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:101]: What's the meaning of integer in `exprToInteger` and `integerToExpr`? Names of variables should be as specific as possible.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java [hunk 8]
```diff
@@ -46,6 +46,15 @@
 public class ExprRewriter {
     private int numChanges_ = 0;
     private final List<ExprRewriteRule> rules_;
+
+    // The type of clause that executes the rule.
+    // This type is only used in InferFiltersRule, other rules are not used
+    public enum ClauseType {
+        ONCLAUSE,
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java:53]:
```suggestion
        ON_CLAUSE,
```

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java [hunk 9]
```diff
@@ -46,6 +46,15 @@
...
+        WHERECLAUSE,
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java:54]:
```suggestion
        WHERE_CLAUSE,
```

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java [hunk 10]
```diff
@@ -46,6 +46,15 @@
...
+        OTHERCLAUSE,    // All other clauses that are not on and not where
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java:55]:
```suggestion
        OTHER_CLAUSE,    // All other clauses that are not on and not where
```

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 11]
```diff
@@ -0,0 +1,523 @@
...
+        }
+    }
+
+    private void buildNewIsNotNullPredicate(Expr expr,
+                                            List<Expr> slotEqSlotExpr,
+                                            Set<Expr> isNullDeDuplication,
+                                            List<Pair<Expr, Boolean>> newExprWithState,
+                                            Analyzer analyzer,
+                                            ExprRewriter.ClauseType clauseType) {
+        if (expr instanceof IsNullPredicate) {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:479]: We should check `IsNullPredicate.isSlotRefChildren` here.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 12]
```diff
@@ -0,0 +1,675 @@
...
+/**
+ * The function of this rule is to derive a new predicate based on the current predicate.
+ * eg.
+ * t1.id = t2.id and t2.id = t3.id and t3.id = 100;
+ * -->
+ * t1.id = 100 and t2.id = 100 and t3.id = 100;
+ *
+ * 1. Register a new rule InferFiltersRule and add it to GlobalState.
+ * 2. Traverse Conjunct to construct on/where equivalence connection, numerical connection and isNullPredicate.
+ * 3. Use Warshall to infer all equivalence connections
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:54]: It's better to add a link to the "Warshll" algorithm, i.e., a link in Wikipedia, which people could refer to for more details.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:165]: Added conditions only for and derivation

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+                for (int index = 1; index < newPredicatesList.size(); index++) {
+                    rewriteExpr = new CompoundPredicate(CompoundPredicate.Operator.AND, rewriteExpr, newPredicatesList.get(index));
+                }
+                return rewriteExpr;
+            }
+        }
+        return expr;
+    }
+
+    private void initAllStructure(Expr conjunct,
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:152]: done

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+            ret = true;
+        }
+        return ret;
+    }
+
+    private Expr genNewBinaryPredicate(Analyzer analyzer, Expr oldExpr, Expr newSlot) {
+        if (oldExpr instanceof BinaryPredicate) {
+            BinaryPredicate oldBP = (BinaryPredicate) oldExpr;
+            BinaryPredicate newBP = new BinaryPredicate(oldBP.getOp(), newSlot, oldBP.getChild(1));
+            newBP.analyzeNoThrow(analyzer);
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java]: deleted

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+                addNewBinaryPredicate(genNewBinaryPredicate(analyzer, slotToLiteral, leftSlot),
+                        slotToLiteralPredicatesSet, newPredicatesList,
+                        tmpSlotToLiteralPredicatesList,
+                        isNeedInfer(checkSlot, leftSlot));
+            }
+        }
+    }
+
+    private boolean isNeedInfer(SlotRef checkSlot, SlotRef newSlot) {
+        JoinOperator jop = newSlot.getDesc().getParent().getRef().getJoinOp();
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java]: Re-adjusted the rules of predicate derivation, considering on_clause and where_clause separately

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java [hunk 16]
```diff
@@ -63,202 +68,307 @@ public Expr apply(Expr expr, Analyzer analyzer) throws AnalysisException {
...
-    private void inferSlotPredicates(ArrayList<Expr> slotEqSlotPredicatesList,
-                                    Set<Pair<Expr, Expr>> slotToLiteralPredicatesSet,
-                                    ArrayList<Expr> slotToLiteralPredicatesList,
-                                    ArrayList<Expr> newPredicatesList,
-                                    Analyzer analyzer) {
-        ArrayList<Expr> tmpSlotToLiteralPredicatesList = new ArrayList<>();
-        for (Expr slotToLiteral : slotToLiteralPredicatesList) {
-            buildNewBinaryPredicate(slotToLiteral, slotEqSlotPredicatesList, slotToLiteralPredicatesSet,
-                    newPredicatesList, tmpSlotToLiteralPredicatesList, analyzer);
+    /**
```
[the remark this answers is not in the document]
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:230]: already edited

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java (hunk 6 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java (hunk 7 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java]: already edited

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java (hunk 8 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java (hunk 9 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java (hunk 10 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java (hunk 11 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java:532]: already edited

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java (hunk 12 above)
Comment (zhengshiJ) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java]: done

**Review by EmmyMiao87 — DISMISSED** [review]
LGTM

**Comment by EmmyMiao87** [comment]
Waiting for UT

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**